Add support for FindsByWindowsAutomationAttribute

cosminstirbu opened the pull request:
Add support for FindsByWindowsAutomationAttribute - see https://github.com/appium/appium-dotnet-driver/issues/237 for more details.

P.S - The name of the branch is a bit misleading since I've rebased it against master (4.0.0.0 beta 4).

akinsolb commented:
@cosminstirbu Are you able to provide some integration tests? in order for it to be merged

cosminstirbu commented:
Unfortunately I'm quite caught up so I won't be able to do it in the near future (2-4 weeks).

If you're OK with it, please merge it as is, and I'll revisit it at some point, either me or someone from my team, as I think it would be a good "exercise" for them to contribute on an open source project.

cosminstirbu commented:
@akinsolb 

I've updated the existing Windows tests because they were failing due to outdated accessibility identifiers.

I've also added an integration tests for the Windows Page Object attributes that I've added.

Please review and let me know if there's anything else that you'd like me to add.

akinsolb commented:
@cosminstirbu I'll try get this done today.  I had build server issues whilst testing last year.

KazuCocoa commented on src/Appium.Net/Appium/Windows/WindowsDriver.cs:
```diff
@@ -110,6 +111,31 @@ public WindowsDriver(Uri remoteAddress, AppiumOptions AppiumOptions, TimeSpan co
+        {
+            get
+            {
+                return "NATIVE_APP";
```
- Does windows driver have no implementation for `get context` as well?
- Does windows driver raise `InvalidOperationException` for `set`? If so, we simply though the error to users.
    - We can reduce client-side implementation

cosminstirbu replied:
WinAppDriver returns Unknown Command as it can be seen here https://gist.github.com/cosminstirbu/14375a1f101ad5fed16648a37b746102

On the client side client, the exception is thrown in `ContentMappedBy` via `GetReturnRelevantBys`. I could do a try / catch there on the catch clause `return map[ContentTypes.NATIVE];`

cosminstirbu replied:
However this would "mess up" the logs since this call will be frequently logged :-?

KazuCocoa replied:
> However this would "mess up" the logs since this call will be frequently logged :-?

It's trade-off thing, I would say.
We are implementing clients to throw errors to users such as `not implemented yet` without modification in clients layer as possible. We can fix the error only updating Appium server side. We do not need to force users to update frequently.

I am not familiar with windriver's situation. Thus, I'm fine to go this way 👍if it is better for windows driver.

cosminstirbu commented on src/Appium.Net/Appium/Windows/WindowsDriver.cs:
```diff
@@ -110,6 +111,31 @@ public WindowsDriver(Uri remoteAddress, AppiumOptions AppiumOptions, TimeSpan co
+
+        public override string Context
+        {
+            get
```
Done

cosminstirbu commented on src/Appium.Net/Appium/Windows/WindowsDriver.cs:
```diff
@@ -110,6 +111,31 @@ public WindowsDriver(Uri remoteAddress, AppiumOptions AppiumOptions, TimeSpan co
+            {
+                return "NATIVE_APP";
+            }
+            set
```
Done

cosminstirbu commented:
@akinsolb I have ran the tests without setting the remoteFlag, I have used only the local service without any other changes (since I have Appium server installed locally, and it comes with WinAppDriver out of the box, so no other changes needed there)

I've copy pasted the setup from the existing tests, so I'm not sure if the existing remote infrastructure is working or not.

akinsolb commented:
Ok. The reason I asked is because the context endpoint isn't supported by the WinApp Driver. I was using the standalone while testing the context, but appium server side handles the getContext endpoint.

All LGTM
